src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -383,8 +383,27 @@ internal bool CheckValueKind(SyntaxNode node, BoundExpression expr, BindValueKin
                     }
                     break;
 
-                // array elements and pointer dereferencing are readwrite variables
+                // array access is readwrite variable if the indexing expression is not System.Range
                 case BoundKind.ArrayAccess:
+                    {
+                        if (RequiresRefAssignableVariable(valueKind))
+                        {
+                            Error(diagnostics, ErrorCode.ERR_RefLocalOrParamExpected, node);
+                            return false;
+                        }
+
+                        var boundAccess = (BoundArrayAccess)expr;
+                        if (boundAccess.Indices.Length == 1 &&
+                            boundAccess.Indices[0].Type.Equals(Compilation.GetWellKnownType(WellKnownType.System_Range)))
```

Comment:
Shouldn't we be using `TypeSymbol.Equals` here?

Reply:
Isn't that what this is? There's a static method that just calls the instance method. Are you worried about if the LHS type is null?

Reply:
I was pretty sure @jcouv was moving everything to be `TypeSymbol.Equals` and removing the instance on entirely. Or at least he's pushing every equality check to use the enum that specifies what type of equality is meant to be done here (null considering or not for example). 

---
In reply to: [248118094](https://github.com/dotnet/roslyn/pull/32449#discussion_r248118094) [](ancestors = 248118094)